teps/0050-ignore-task-failures.md

```diff
@@ -187,6 +194,220 @@ control over the task definitions but may desire to ignore a failure and continu
+}
+```
+
+```go
+type OnErrorType string
+
+const (
+	// StopAndFail indicates to stop the pipeline if the task is failed
+	StopAndFail OnErrorType = "stopAndFail"
+	// Continue indicates to continue executing the rest of the pipeline irrespective of the status of the task
+	Continue    OnErrorType = "continue"
+)
+```
+
+Pipeline author can set the ```OnError``` field to configure the task failure strategy. If set to ```StopAndFail```, the pipeline is stopped and failed when the task is failed. If set to ```Continue```, the failure of task is ignored and the pipeline continues to execute the rest of the DAG.
+
+```yaml
+- name: task1
+  onError: continue
+  taskSpec:
+    steps:
+      - image: alpine
+        name: exit-with-1
+        script: |
+          exit 1
+
+```
+This new field ```OnError``` will be implemented as a alpha feature and can be enabled by setting enable-api-fields to alpha.
+
+Setting ```OnError``` is optional, the default pipeline behaviour is ```StopAndFail```
+
+The task run information is avaialble under the ```pipelineRun.status.taskRuns```. Note that the original task run status remains as it is irrelevant of the value of ```OnError``` (i.e. a failed task with ```onError: continue``` is still marked as failed). The task would be considered "successful" ONLY for the purposes of determining the status of the pipeline, which is represented in ```pipelineRun.status.conditions``` 
+
+
+To distinguish pipeline run messages with and without ignored task failures, we explicitly add the ignored task failure count to ```pipelineRun.status.conditions.message``` in the following way if ignored task failure > 0: 
+
+```
+"Tasks Completed: A (Failed: B (C is ignored), Cancelled D), Skipped: E"
+```
+
+Example Input:
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata: 
+  name: demo-pipeline-run
+spec:
+  pipelineSpec:
+    tasks:
+    - name: task1
+      onError: continue
+      taskSpec:
+        steps:
+          - image: alpine
+            name: exit-with-1
+            script: |
+              exit 1
+    - name: task2
+      taskSpec:
+        steps:
+          - image: alpine
+            name: exit-with-0
+            script: |
+              exit 0
+```
+
+Example Output:
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+...
+status:
+  completionTime: "2022-08-15T17:26:15Z"
+  conditions:
+    - lastTransitionTime: "2022-08-15T17:26:15Z"
+      message: "Tasks Completed: 2 (Failed: 1 (1 is ignored), Cancelled 0), Skipped: 0"  
+      reason: Succeeded
+      status: "True"  # The failed task is considered "successful" when determining the state of pipelineRun
+      type: Succeeded
+  pipelineSpec:
+    ...
+  taskRuns:
+    demo-pipeline-run-task1:
+      pipelineTaskName: task1
+      status:
+        completionTime: "2022-08-15T17:26:13Z"
+        conditions:
+          - lastTransitionTime: "2022-08-15T17:26:13Z"
+            message: ...
```

Comment:
can we add some info in `message` that the failure happened but it was ignored?

Reply:
Good point. Added a prefix [Task is failed but ignored] to the```message``` section